package-recipe.el

```diff
@@ -0,0 +1,108 @@
+;;; package-recipe.el --- Package recipes as EIEIO objects  -*- lexical-binding: t -*-
+
+;; Copyright (C) 2018  Jonas Bernoulli
+
+;; Author: Jonas Bernoulli <jonas@bernoul.li>
+
+;; This file is not (yet) part of GNU Emacs.
+;; However, it is distributed under the same license.
+
+;; GNU Emacs is free software; you can redistribute it and/or modify
+;; it under the terms of the GNU General Public License as published by
+;; the Free Software Foundation; either version 3, or (at your option)
+;; any later version.
+
+;; GNU Emacs is distributed in the hope that it will be useful,
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+
+;; You should have received a copy of the GNU General Public License
+;; along with GNU Emacs; see the file COPYING.  If not, write to the
+;; Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+;; Boston, MA 02110-1301, USA.
+
+;;; Commentary:
+
+;; Package recipes as EIEIO objects.
+
+;;; Code:
+
+(require 'eieio)
+
+(declare-function package-build-recipe-alist "package-build" ())
+(defvar package-build-working-dir)
+
+;;; Classes
+
+(defclass package-recipe ()
+  ((url-format      :allocation :class       :initform nil)
```

Comment:
Are all the `:initform nil` specifications really necessary? One would assume that `nil` would be the default anyway.

Reply:
They are, in fact, necessary.

```emacs-lisp
(defclass foo ()
  ((foo-field)))

(oref (make-instance 'foo) foo-field)
;; => Debugger entered--Lisp error: (unbound-slot foo "#<foo foo>" foo-field oref)
```

Reply:
Just because it is an error to access an unbound slot doesn't mean that we have to make sure it is never unbound. I have just decided that prevent that kind of error would be a good idea during this stage of development.